lib/main.ts

```diff
@@ -52,636 +52,565 @@ const SECRETS = JSON.parse(shell('sh ./generate_secrets.sh').toString());
-                if (expansions[i].length === 2) {  // !! expansion
-                    expansions[i] = await this.last_message({
-                        command: true,
-                        channel: message.channel
-                    });
-                    continue;
-                }
-
-                const [opts, offset] = expansions[i].slice(2)
-                    .replace(/(\d+)/, '#$1')
-                    .split('#');
-
-                const mention = opts.includes('@');
-                const mentioning = expansions[i + 2];
-                if (mention) expansions[i + 2] = '';
-
-                expansions[i] = await this.last_message({
-                    command: !opts.includes('^'),
-                    mention: mention,
-                    mentioning: mentioning,
-                    offset: offset || 1,
-                    channel: message.channel
-                });
-            }
-        }
-
-        // TODO: Deal with 'EXPANSION_ERROR's in `expansions`.
-
-        return expansions.join('');
-    }
-
-    @On("message")
-    async on_message(message : Message, client : Client) {
-        console.log('Message acknowledged.');
-        if (SimpOMatic._client.user.id === message.author.id) {
-            return;
-        }
-        console.log('Message received:', message.content);
-
-        const trimmed = message.content.trim();
-        message.content = trimmed;
-        // When finished expanding...
-        this.expand(message).then(content => {
-            if (content.length >= 2000) {
-                message.answer("The expansion for that message was"
-                + " over 2000 characters, what the fuck is wrong with you?");
-                return;
-            }
-            message.content = content;
-            console.log('Expanded message:', message.content);
-
-            if (message.content[0] === CONFIG.commands.prefix) {
-                console.log('Message type: command.')
-                this.process_command(message);
-            } else {
-                console.log('Message type: generic.')
-                this.process_generic(message);
-            }
-        });
-    }
+	private static _client : Client;
+	private _COMMAND_HISTORY : Message[] = [];
+
+	constructor() {
+		console.log('Secrets:', pp(SECRETS));
+		console.log('Configured Variables:', pp(CONFIG));
+		console.log('Known commands:', pp(KNOWN_COMMANDS));
+	}
+
+	static start() {
+		this._client = new Client();
+		this._client.login(
+			SECRETS.api.token,
+			`${__dirname}/*Discord.ts`
+		);
+	}
+
+	process_command(message : Message) {
+		const last_command = this._COMMAND_HISTORY.last();
+		this._COMMAND_HISTORY.push(message);
+		if (this._COMMAND_HISTORY.length > CONFIG.commands.max_history) {
+			this._COMMAND_HISTORY.shift();
+		}
+		const current_command = this._COMMAND_HISTORY.last();
+
+		// Try and slow the fellas down a little.
+		if (last_command.channel === current_command.channel) {
```

Comment:
This line and a couple below it were changed a few minutes ago, you wanna pull that quick?

Reply:
No idk how to undo it